### ISSUE TITLE
Diagnostic clarity: misleading error message

### ISSUE THREAD

**Issue by controlflow** [dotnet/roslyn#17254]
Error messages for constant patterns under `is` can be misleading comparing to similar situation with `switch` statement patterns:
```c#
class C {
  public const int X = 42;
}

class D {
  public const string C = "abc";

  // error CS0426: The type name 'X' does not exist in the type 'C'
  public bool M1(object o) => o is C.X;
    
  public bool M2(object o) {
      switch (o) {
          // error CS1061: 'string' does not contain a definition for 'X' and no extension
          // method 'X' accepting a first argument of type 'string' could be found
          // (are you missing a using directive or an assembly reference?)
          case C.X:
              return true;
          default:
              return false;
      }
  }
}
```

**Comment by gafter** [comment]
The message is different because the meaning is different.

`e is A.B` looks for a type named `A.B` as its first choice for how to interpret this. That is required for backward compatibility.

`case A.B` looks for an expression named `A.B` (which is hoped to be a constant). That too is required for backward compatibility.

That is the reason the errors are different.

Does this answer address your concern/question? What would you prefer to be different?

**Comment by alrz** [comment]
>That is required for backward compatibility.

Could that be relaxed to allow those cases but with a precedence for binding according to previous versions to retain backward compatibility? e.g. `e is X` first look for a type and then a constant, conversely, `case X` look for a constant and then for a type?

**Comment by gafter** [comment]
@alrz That is exactly what we already do.

**Comment by controlflow** [comment]
@gafter the error states that "The type name 'X' does not exist in the type 'C'", so it looks like the `C` reference in `o is C.X` is bound to type `C`, but in fact it is bound to constant `D.C`. Removing `D.C` constant fixes the error.

**Comment by alrz** [comment]
@gafter Somehow, but we don't go further in case of a failure. If we were, the example in the OP would compile fine.  `C.X` as a type is unresolvable in `is C.X`, we could, however, bind it to the constant `C.X`, or vice versa for `case C.X`.

**Comment by gafter** [comment]
@controlflow The name `C.X` must be bound either as a type or as a (constant) expression. As an expression it fails (because there is no constant named `X` inside `D.C`, which is what `C` as an expression binds to). As a type it fails because there is no such type `C.X` anywhere. Since the backward-compatible behavior is to treat it as a type, we select which of these two errors to report by doing so in a way analogous to the way the previous compiler did it, by reporting the error when looking up the type.

**Comment by controlflow** [comment]
@gafter the only confusing part is "does not exist in the type 'C'". If expression `C` is bound to `D.C`, it has type `string`, not `C`, isn't it?

**Comment by gafter** [comment]
@controlflow When we're looking up a type, we don't bind the part before the dot as an expression.

**Comment by controlflow** [comment]
OK, I think I finally got it 👍
